libs/hydrateAndFilter.js

```diff
@@ -1,45 +1,63 @@
 /**
  * Used not to duplicate data.
  *
- * * `options` :
- *    - `keys` : List of keys used to check that two items are the same. By default it is set to `['id']'.
- *    - `index` : Return value returned by `cozy.data.defineIndex`, the default will correspond to all documents of the selected doctype.
- *    - `selector` : Mango request to get records. Default is built from the keys `{selector: {_id: {"$gt": null}}}` to get all the records.
+ * `options`:
+ * - `index` : As returned by `cozy.data.defineIndex`. Default corresponds
+ *   to all documents of the selected doctype
+ * - `selector` : Mango query. Default one is `{selector: {_id: {"$gt": null}}}` to
+ *    get all the records.
+ * - `keys` : List of keys used to check that two items are the same. Default is `['_id']`
  *
  * @module filterData
  */
 
 const bluebird = require('bluebird')
 const log = require('./logger').namespace('hydrateAndFilter')
+const get = require('lodash/get')
+const uniqBy = require('lodash/uniqBy')
+
+/**
+ * Since we can use methods or basic functions for
+ * `shouldSave` and `shouldUpdate` we pass the
+ * appropriate `this` and `arguments`.
+ *
+ * If `funcOrMethod` is a method, it will be called
+ * with args[0] as `this` and the rest as `arguments`
+ * Otherwise, `this` will be null and `args` will be passed
+ * as `arguments`.
+ */
+const suitableCall = (funcOrMethod, ...args) => {
+  const arity = funcOrMethod.length
+  if (arity < args.length) {
+    // must be a method
+    return funcOrMethod.apply(args[0], args.slice(1))
+  } else {
+    // must be a function
+    return funcOrMethod.apply(null, args)
+  }
+}
 
 const hydrateAndFilter = (entries, doctype, options = {}) => {
   const cozy = require('./cozyclient')
 
   log('debug', String(entries.length), 'Number of items before hydrateAndFilter')
   if (!doctype) return Promise.reject(new Error(`Doctype is mandatory to filter the connector data.`))
 
-  // expected options:
-  //  - index : this is return value which returned by cozy.data.defineIndex, the default will
-  //  correspond to all document of the selected doctype
-  //  - selector : this the mango request : default one will be {selector: {_id: {"$gt": null}}} to
-  //  get all the records
-  //  - keys : this is the list of keys used to check that two items are the same
   const keys = options.keys ? options.keys : ['_id']
   const store = {}
 
   log('debug', keys, 'keys')
 
   const createHash = item => {
     return keys.map(key => {
-      let result = item[key]
+      let result = get(item, key)
```

Comment:
Why is lodash/get needed here ? I find it less readable like this.

Reply:
Because you may want to use a path instead of the name of an attribute.

`keys = ["metadata.version", "name"]` would work here.